### PR TITLE
Deprecate registering autograd kernels at not an autograd key

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #104481

Context
-------
This PR adds a new fallback to the Autograd dispatch keys.

If you would prefer the old behavior:
- A quick (unsupported) way to get the previous behavior is to call
`torch._C._set_autograd_fallback("nothing")`
- Register "torch::CppFunction::makeFallthrough()" to your Autograd key,
like in https://gist.github.com/zou3519/d09a5f4b1afe2430af09fea67c6ff2c8

It is possible that this PR regresses performance of overhead-bound
models. If this is the case, please reach out (and apply one of the
temporary fixes in the previous section).

Description for reviewers
-------------------------
In order to deprecate registering autograd kernels at not an autograd
key, we add a fallback to the Autograd dispatch keys. This fallback
raises a warning if the user attempts to backprop through the operator
and is also configurable to either warn or not warn.

The goal of this PR is to
- preserve as much BC as possible
- raise a warning that whatever the user is doing is potentially wrong.
- be as performant as possible

There are roughly two cases:
- if the post-autograd kernels return a Tensor that requires grad, then
we install an autograd hook that raises a warning. We are preserving BC
in that it is possible that the user has a torch::autograd::Function
registered to their CPU key.
- if the post-autograd kernels return Tensors that do not require grad,
then we make them require_grad and install a WarnNotImplemented grad fn
that warns in the backward pass. This is mildy BC-breaking (see next
section).

Test Plan:
- bunch of new tests

BC-Breaking Note
----------------
This PR adds a new fallback to the Autograd dispatch keys. It affects
custom operators that do not have a kernel registered to the Autograd
keys (e.g. AutogradCPU and AutogradCUDA).

If the previous behavior was that the custom operator would return
Tensors that do not require grad if the inputs do require grad, then
this PR changes it so that all floating-point and complex returns do
require grad. See the "Context" section above for how to get the old
behavior.

cc @ezyang @gchanan

### CODE REVIEW

**Comment by zou3519** [comment]
Not ready for review yet; I want to see CI signals

**Comment by zou3519** [comment]
Should be ready for review now

**Review by albanD — COMMENTED** [review]
Sounds good to me. I'll let @soulitzer do a detailed review

File: test/autograd/test_fallback.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+
+        lib.impl("foo", foo_impl, "CPU")
+        x = torch.randn(3, requires_grad=True)
+        y = torch.randn(1, requires_grad=True)
+        z = torch.randn(2, 1, requires_grad=True)
+        a, b = op([x, y, z])
+        with self._check_ctx(mode):
+            torch.autograd.grad(a, (x, y, z), torch.ones_like(a), allow_unused=True, retain_graph=True)
+        with self._check_ctx(mode):
+            torch.autograd.grad(b, (x, y, z), torch.ones_like(b), allow_unused=True, retain_graph=True)
```
Comment (soulitzer) [inline test/autograd/test_fallback.py:296]: In the mode="nothing" case, I'd expect the returned outputs `a, b` to not require grad and result in an error when doing `.grad` right?

File: test/autograd/test_fallback.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+        x, y, z = op(a, b)
+
+        with self._check_ctx(mode):
+            torch.autograd.grad(x, (a, b), torch.ones_like(x), allow_unused=True, retain_graph=True)
+
+        with self._check_ctx(mode):
+            torch.autograd.grad(y, (a, b), torch.ones_like(y), allow_unused=True, retain_graph=True)
+
+        with self._check_ctx(mode):
+            torch.autograd.grad(z, (a, b), torch.ones_like(z), allow_unused=True, retain_graph=True)
```
Comment (soulitzer) [inline test/autograd/test_fallback.py:275]: Same here (see below), I'd expect the `x` and `z` ones to error in the mode="nothing" case.

File: test/autograd/test_fallback.py (hunk 1 above)
Comment (zou3519) [reply test/autograd/test_fallback.py:296]: Oh that's a good catch. I forgot to set the mode in many of these tests, will update.

File: torch/csrc/autograd/autograd_not_implemented_fallback.h [hunk 3]
```diff
@@ -5,9 +5,28 @@
 namespace torch {
 namespace autograd {
 
+// DispatchKey::Autograd fallback for built-in operators
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/csrc/autograd/autograd_not_implemented_fallback.h:8]: No, `autogradNotImplementedFallback()` is stricter than `basicAutogradNotImplementedFallback` in that it errors out, so we would prefer users to use `autogradNotImplementedFallback`.

File: torch/csrc/autograd/autograd_not_implemented_fallback.cpp [hunk 4]
```diff
@@ -45,8 +45,170 @@ void _foreach_tensor(
...
+            return;
+          }
+
+          // If the post-autograd implementation returns any Tensors that
+          // don't require grad, then we install the WarnNotImplemented grad_fn.
+          // This grad_fn warns in backward and returns undefined tensor
+          // gradients.
+          if (is_mutable_output) {
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+            rebase_history(const_cast<at::Tensor&>(t), grad_fn);
```
Comment (soulitzer) [inline torch/csrc/autograd/autograd_not_implemented_fallback.cpp:195]: I suppose we would reach here if the user passed in a view that doesn't require grad as input and return that input as-is. (User may have also done inplace ops, and if those operations involved any other inputs that requires grad it would've necessarily been done in no-grad).

I think we should just not do anything in this case, rather than rebase_history. If we're doing rebase_history, we're making the view and base which previously did not require grad, now require grad. I can see that being potentially problematic if the user uses the base/view in subsequent computation not expecting that it now requires grad and saving some saved tensors unnecessarily.

File: torch/csrc/autograd/autograd_not_implemented_fallback.cpp [hunk 5]
```diff
@@ -45,8 +45,170 @@ void _foreach_tensor(
...
+            // users typically call .backward() and backprop through
+            // the entire program).
+            if (t.is_view() && is_mutable_output) {
+              // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+              const_cast<at::TensorBase&>(t._base()).register_hook(
+                  [op_name](const at::TensorBase& grad) {
+                    warnAutogradNotImplemented(op_name);
+                    return grad;
+                  });
+            }
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/csrc/autograd/autograd_not_implemented_fallback.cpp:185]: Ah I think I misread, thanks for clarifying

File: torch/csrc/autograd/autograd_not_implemented_fallback.cpp [hunk 6]
```diff
@@ -45,8 +45,170 @@ void _foreach_tensor(
...
+  TORCH_WARN(
+      op_name,
+      ": an autograd kernel was not registered to the Autograd key(s) ",
+      "but we are trying to backprop through it. This may lead to silently incorrect behavior. ",
+      "This behavior is deprecated and will be removed in a future version of PyTorch. ",
+      "If your operator is differentiable, please ensure you have registered an "
+      "autograd kernel to the correct Autograd key (e.g. DispatchKey::Autograd, "
+      "DispatchKey::CompositeImplicitAutograd). If your operator is not "
+      "differentiable, or to squash this warning and use the previous behavior, "
+      "please register torch::CppFunction::makeFallthrough() to DispatchKey::Autograd.");
```
Comment (soulitzer) [inline torch/csrc/autograd/autograd_not_implemented_fallback.cpp:71]: Is it worth specifically pointing out interactions with Python key? For example if I trace with LoggingMode, that can result in the autograd key being excluded and changing the results of my computation.

File: torch/csrc/autograd/autograd_not_implemented_fallback.cpp (hunk 4 above)
Comment (zou3519) [reply torch/csrc/autograd/autograd_not_implemented_fallback.cpp:195]: Note that the user would have also marked that particular output as being written to in the schema, e.g. `(Tensor(a!) x) -> Tensor(a!)`. We can avoid the rebase_history to be on the safe side (to be less BC-breaking), it just comes at a tradeoff of not being able to catch situations where the user does do an in-place operation on an input view and we aren't able to warn them that they didn't add an autograd kernel

File: torch/csrc/autograd/autograd_not_implemented_fallback.cpp (hunk 6 above)
Comment (zou3519) [reply torch/csrc/autograd/autograd_not_implemented_fallback.cpp:71]: >  For example if I trace with LoggingMode, that can result in the autograd key being excluded and changing the results of my computation.

What do you mean? I'm not familiar with how LoggingMode looks like anymore

File: torch/csrc/autograd/autograd_not_implemented_fallback.cpp (hunk 4 above)
Comment (soulitzer) [reply torch/csrc/autograd/autograd_not_implemented_fallback.cpp:195]: If they had other returns, then they should still get a warning when doing backward. I guess we would no longer have a warning for the case where the user's only return was the in-place one. 
The only case where that output would naturally still be requires_grad=False is any computation involving other inputs that require grad were done in no-grad mode - in that case it seems that the users intention is to have the kernel be non-differentiable.

File: torch/csrc/autograd/autograd_not_implemented_fallback.cpp (hunk 6 above)
Comment (soulitzer) [reply torch/csrc/autograd/autograd_not_implemented_fallback.cpp:71]:
```python
from torch.library import Library
from torch.testing._internal.logging_tensor import capture_logs_with_logging_tensor_mode

test_ns = "abc"
lib = Library(test_ns, "FRAGMENT")
lib.define("foo(Tensor(a!) a) -> Tensor(a!)")

def get_op(name):
    return getattr(getattr(torch.ops, test_ns), name).default

op = get_op("foo")

# User initially defines this as a composite op
def foo_impl(a):
    a *= 2
    return a

lib.impl("foo", foo_impl, "CPU")
x = torch.randn(3, requires_grad=True)
x_clone = x.clone()

with capture_logs_with_logging_tensor_mode() as logs:
  y = op(x_clone)

dx, = torch.autograd.grad(y.sum(), x)
print(dx)
```
Normally, autograd key is a fallthrough, so when we run foo_impl everything is OK, we build the graph as expected and the gradient is correct `tensor([2., 2., 2.])`
But, if logging_mode is enabled, then dx returns a tensor full of ones because foo_impl would be run with autograd key disabled.

Comment (zou3519) [reply torch/csrc/autograd/autograd_not_implemented_fallback.cpp:71]: I see, thanks for the explanation.

I'd prefer to skip an explanation of the interaction with the Python dispatch key to keep the error message simple; also,  I agree that the above behavior falls under "this may lead to silently incorrect behavior"

File: torch/csrc/autograd/autograd_not_implemented_fallback.cpp (hunk 4 above)
Comment (soulitzer) [reply torch/csrc/autograd/autograd_not_implemented_fallback.cpp:195]: Actually the real reason to not do this is because rebase_history on a tensor that doesn't require grad is undefined behavior. 
You end up having a graph that doesn't even end with a leaf node!

Comment (soulitzer) [reply torch/csrc/autograd/autograd_not_implemented_fallback.cpp:195]: I think we're missing an assert somewhere

File: torch/csrc/autograd/autograd_not_implemented_fallback.cpp (hunk 6 above)
Comment (soulitzer) [reply torch/csrc/autograd/autograd_not_implemented_fallback.cpp:71]: sounds good!

File: torch/csrc/autograd/autograd_not_implemented_fallback.cpp (hunk 4 above)
Comment (zou3519) [reply torch/csrc/autograd/autograd_not_implemented_fallback.cpp:195]: > Actually the real reason to not do this is because rebase_history on a tensor that doesn't require grad should be considered undefined behavior.
You end up having a graph that doesn't even end with a leaf node!

Fair enough. When we get to this point the Tensor doesn't require grad :P.
So, I'll delete the rebase_history and we will live with some code that doesn't catch all of the cases.

Comment (soulitzer) [reply torch/csrc/autograd/autograd_not_implemented_fallback.cpp]: > Actually the real reason to not do this is because rebase_history on a tensor that doesn't require grad should be considered undefined behavior.
You end up having a graph that doesn't even end with a leaf node!

I looked into this a bit more, and I realize what the problem is now.

So both of the following are fine without using custom ops. So clearly it is supposed to work. (sorry I should've checked these cases earlier)
```
# Rebasing on tensor that does not require grad
a = torch.tensor(1.)
b = torch.tensor(1., requires_grad=True)
a += b

# Rebasing on a view that does not grad, and whose base does not require grad
a = torch.tensor([1.])[:]
b = torch.tensor(1., requires_grad=True)
a += b
print(a._base.grad_fn.next_functions)
```

The key difference between what this fallback does and what a VariableType kernel normally does it that normally a VariableType kernel collects next edges from ALL of the inputs, including the ones that do not require grad, versus in this PR, we only collect next edges from inputs that do require grad. Normally this is fine, but the difference matters in the specific case where we do inplace over view, where we need to create a CopySlices node.

Taking a look at the CopySlices node, it notably skips over the first edge in the grad_fn its supposed to wrap, i.e. the one corresponding to the tensor being inplaced. This means that it is actually necessary to have collected the edge from the first input even though it does not require grad as to accomodate this skip.

Comment (zou3519) [reply torch/csrc/autograd/autograd_not_implemented_fallback.cpp]: Resolved offline:
- if an output is mutated according to the schema and doesn't require grad, we will not do anything (rebase or set history).
- We can improve this case in the future, but right now we are OK with the tradeoff.

File: test/autograd/test_fallback.py [hunk 7]
```diff
@@ -241,6 +242,131 @@ def backward(ctx, gx):
...
+
+            lib.impl("foo", foo_impl, "CPU")
+            op = self.get_op("foo")
+
+            x = torch.randn(3)
+            y = torch.randn(3, requires_grad=True)
+            with self.assertRaisesRegex(RuntimeError, "does not require grad"):
+                z = x.clone()
+                op(z, y)
+                torch.autograd.grad(z, y, torch.ones_like(z), allow_unused=True)
```
Comment (soulitzer) [inline test/autograd/test_fallback.py:269]: nit:
- we could probably deduplicate everything except the function definition since they are all called the same way
- we could also move everything except the grad/backward call from under assertRaises 

(same below)

File: torch/csrc/autograd/autograd_not_implemented_fallback.cpp [hunk 8]
```diff
@@ -113,22 +113,31 @@ static void basicAutogradNotImplementedFallbackImpl(
...
+    std::vector<const at::Tensor*> tensors_requiring_grad_on_stack;
+    _foreach_tensor(
+        [&](size_t _, size_t idx_arg, const at::Tensor& t) {
+          if (grad_mode && t.requires_grad()) {
+            tensors_requiring_grad_on_stack.push_back(&t);
+          }
+        },
+        stack,
+        stack_start,
+        num_arguments);
```
Comment (soulitzer) [inline torch/csrc/autograd/autograd_not_implemented_fallback.cpp:140]: Can we add a comment that this differs from what VariableType is doing? e.g. in VariableType we gather next edges from ALL tensors even if that tensor does not require grad.
Or, just changing it to be the same as what VariableType might be better (though I don't know what breaks) - and a quick comment in that case would still be good, since this is kind of unexpected, like you could mention the CopySilces thing.

File: test/autograd/test_fallback.py (hunk 7 above)
Comment (soulitzer) [reply test/autograd/test_fallback.py:269]: We could also combine the two tests if you'd like

**Review by soulitzer — APPROVED** [review]
LGTM!, just small nits

File: torch/csrc/autograd/autograd_not_implemented_fallback.cpp (hunk 8 above)
Comment (zou3519) [reply torch/csrc/autograd/autograd_not_implemented_fallback.cpp:140]: We could change it to what VariableType does. This PR is modeled after the existing AutogradNotImpelementedFallback, which only gathers the edges that requires grad.

I tried this out locally before (gather next edges from all tensors), and nothing breaks, but I'll poke at it some more

**Comment by kumpera** [comment]
The gloo failure looks unrelated as I can reproduce it on main. We can ignore the UCC one as well.

**Comment by zou3519** [comment]
https://github.com/pytorch/pytorch/pull/104961 is a version of this PR that doesn't actually change the fallback (but does add the functions in C++), but the above two tests fail. I think those two tests are just flaky
